File: ios/agora-trulience-sdk/ContentView.swift
//
//  ContentView.swift
//  agora-trulience-sdk
//
//  Created by Mateusz on 08/04/2025.
//

import SwiftUI

struct ConnectionInfo {
    var appId: String = ""
    var channelName: String = ""
    var uid: String = ""
    var avatarId: String = ""
    var voiceId: String = ""
    var prompt: String = ""
    var greeting: String = ""
}

struct ContentView: View {
    @State private var connectionInfo = ConnectionInfo(
        appId: "20b7c51ff4c644ab80cf5a4e646b0537",
        channelName: "random",
        uid: "111",
        avatarId: "3384296204170052843"
    )
    
    @State private var navigateToWebView = false
    
    var body: some View {
        NavigationView {
            ZStack {
                VStack(spacing: 30) {
                    // Title
                    Text("Agora convoAI and\nTrulience Avatar Demo")
                        .font(.title)
                        .fontWeight(.bold)
                        .foregroundColor(.white)
                        .padding(.top, 20)
                        .multilineTextAlignment(.center)
                    
                    // Connection Details Card
                    VStack(alignment: .leading, spacing: 15) {
                        Text("Agora Details")
                            .font(.headline)
                            .foregroundColor(.white)
                        
                        // Text Fields
                        TextField("App ID", text: $connectionInfo.appId)
                            .textFieldStyle(RoundedBorderTextFieldStyle())
                            .autocapitalization(.none)
                        
                        TextField("Channel Name", text: $connectionInfo.channelName)
                            .textFieldStyle(RoundedBorderTextFieldStyle())
                            .autocapitalization(.none)
                        
                        TextField("UID", text: $connectionInfo.uid)
                            .textFieldStyle(RoundedBorderTextFieldStyle())
                            .keyboardType(.numberPad)
                        
                        TextField("Avatar ID", text: $connectionInfo.avatarId)
                            .textFieldStyle(RoundedBorderTextFieldStyle())
                            .keyboardType(.numberPad)
                        
                        TextField("Voice ID", text: $connectionInfo.voiceId)
                            .textFieldStyle(RoundedBorderTextFieldStyle())
                            .autocapitalization(.none)
                        
                        TextField("Prompt", text: $connectionInfo.prompt)
                            .textFieldStyle(RoundedBorderTextFieldStyle())
                            .autocapitalization(.none)
                        
                        TextField("Greeting", text: $connectionInfo.greeting)
                            .textFieldStyle(RoundedBorderTextFieldStyle())
                            .autocapitalization(.none)
                    }
                    .padding()
                    .background(Color.white.opacity(0.2))
                    .cornerRadius(15)
                    .padding(.horizontal)
                    
                    // Connect Button inside NavigationLink
                    NavigationLink(
                        destination: LazyView(
                            WebView(
                                url: Bundle.main.url(
                                    forResource: "index",
                                    withExtension: "html",
                                    subdirectory: "build")!,
                                connectionInfo: connectionInfo
                            )
                            .navigationBarHidden(true)
                            .edgesIgnoringSafeArea([.bottom])
                            .background(Color.black) 
                        ),
                        isActive: $navigateToWebView,
                        label: {
                            // Button styling
                            Button(action: {
                                navigateToWebView = true
                            }, label: {
                                Text("Connect")
                                    .font(.headline)
                                    .foregroundColor(.white)
                                    .padding()
                                    .frame(maxWidth: .infinity)
                                    .background(
                                        .purple
                                    )
                                    .cornerRadius(10)
                                    .shadow(color: .black.opacity(0.3), radius: 5, x: 0, y: 5)
                            })
                            .padding(.horizontal)
                        }
                    )
                    .disabled(connectionInfo.appId.isEmpty ||
                              connectionInfo.channelName.isEmpty ||
                              connectionInfo.uid.isEmpty)
                    
                    Spacer()
                }
                .padding()
            }.background(.blue)
            .navigationBarHidden(true)
        }
    }
}

#Preview {
    ContentView()
}

File: android/app/src/main/java/com/example/agora_trulience_sdk/ContentView.kt
package com.example.agora_trulience_sdk

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import androidx.navigation.compose.rememberNavController
import java.io.Serializable

data class ConnectionInfo(
    var appId: String = "",
    var channelName: String = "",
    var uid: String = "",
    var voiceId: String = "",
    var prompt: String = "",
    var greeting: String = "",
    var avatarId: String = ""
) : Serializable

/**
 * ContentView composable displays a UI for entering Agora connection details.
 */

@Composable
fun ContentView(navController: NavController) {
    var connectionInfo by remember { mutableStateOf(ConnectionInfo(
        appId= "20b7c51ff4c644ab80cf5a4e646b0537",
        channelName= "random",
        uid= "111",
        avatarId= "9053143346212585739",
        voiceId = "",
        prompt = "",
        greeting = ""
    )) }

    val scrollState = rememberScrollState()

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(color = Color(0xFF007AFF))
            .padding(20.dp)
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(scrollState)
                .imePadding(), // Ensures padding when keyboard is shown
            verticalArrangement = Arrangement.Top,
        ) {
            Spacer(modifier = Modifier.height(24.dp))

            Text(
                text = "Agora convoAI and Trulience Avatar Demo",
                fontSize = 24.sp,
                fontWeight = FontWeight.Bold,
                color = Color.White,
                textAlign = TextAlign.Center,
                modifier = Modifier
                    .padding(bottom = 16.dp, top = 8.dp)
                    .align(Alignment.CenterHorizontally)
                    .width(250.dp)
            )

            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .background(color= Color(0x3395FFFF), shape = RoundedCornerShape(15.dp))
                    .padding(16.dp)
            ) {
                Column(modifier = Modifier.padding(16.dp)) {
                    Text(
                        text = "Agora Details",
                        style = MaterialTheme.typography.titleLarge,
                        color = Color.White
                    )

                    Spacer(modifier = Modifier.height(8.dp))

                    CustomTextField("App ID", connectionInfo.appId) {
                        connectionInfo = connectionInfo.copy(appId = it)
                    }

                    CustomTextField("Channel Name", connectionInfo.channelName, KeyboardType.Number) {
                        connectionInfo = connectionInfo.copy(channelName = it)
                    }

                    CustomTextField("UID", connectionInfo.uid, KeyboardType.Number) {
                        connectionInfo = connectionInfo.copy(uid = it)
                    }

                    CustomTextField("Avatar ID", connectionInfo.avatarId) {
                        connectionInfo = connectionInfo.copy(avatarId = it)
                    }

                    CustomTextField("Voice ID", connectionInfo.voiceId) {
                        connectionInfo = connectionInfo.copy(voiceId = it)
                    }

                    CustomTextField("Prompt", connectionInfo.prompt) {
                        connectionInfo = connectionInfo.copy(prompt = it)
                    }

                    CustomTextField("Greeting", connectionInfo.greeting) {
                        connectionInfo = connectionInfo.copy(greeting = it)
                    }
                }
            }

            Spacer(modifier = Modifier.height(24.dp))

            Button(
                onClick = {
                    navController.currentBackStackEntry?.savedStateHandle?.set("connectionInfo", connectionInfo)
                    navController.navigate(Screen.WebView.route)
                },
                modifier = Modifier
                    .fillMaxWidth()
                    .height(50.dp)
                    .padding(horizontal = 4.dp),
                contentPadding = PaddingValues(),
                colors = ButtonDefaults.buttonColors(
                    containerColor = Color(0xffaf52de),
                    disabledContainerColor = Color.Gray
                ),
                shape = RoundedCornerShape(10.dp),
                elevation = ButtonDefaults.buttonElevation(5.dp)
            ) {
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(
                           color = Color(0xffaf52de),
                            shape = RoundedCornerShape(10.dp)
                        ),
                    contentAlignment = Alignment.Center
                ) {
                    Text(
                        text = "Connect",
                        fontWeight = FontWeight.SemiBold,
                        color = Color.White
                    )
                }
            }

            Spacer(modifier = Modifier.height(24.dp))
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CustomTextField(
    label: String,
    value: String,
    keyboardType: KeyboardType = KeyboardType.Text,
    onValueChange: (String) -> Unit
) {
    OutlinedTextField(
        value = value,
        onValueChange = onValueChange,
        label = { Text(label, color = Color.White) },
        textStyle = TextStyle(color = Color.White),

        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 4.dp),
        keyboardOptions = KeyboardOptions(keyboardType = keyboardType),
        colors = TextFieldDefaults.outlinedTextFieldColors(
            focusedBorderColor = Color.White,
            unfocusedBorderColor = Color.White,
            cursorColor = Color.White,
            focusedTextColor = Color.White,
            unfocusedTextColor = Color.White
        )
    )
}


@Preview(showBackground = true)
@Composable
fun PreviewContentView() {
    ContentView(navController = rememberNavController())
}